engine/runtime/src/main/java/org/enso/interpreter/node/callable/argument/ReadArgumentCheckNode.java

```diff
@@ -254,6 +256,16 @@ Object doPanicSentinel(VirtualFrame frame, PanicSentinel panicSentinel) {
       throw panicSentinel;
     }
 
+    @Specialization
+    Object doUnresolvedConstructor(
```

Comment:
Most of the support for _lazy constructors_ is hidden in this very simple `doUnresolvedConstructor` method. Such a small change is only possible because it **reuses logic** of already implemented **runtime type checking**. Turning `UnresolvedConstructor` into real constructor of the _requested type_ is similar to applying a `from` conversion from one type to the _requested type_.

As such I claim: the logic for _autoscoped constructors_ **must be the same** as the logic for _runtime conversions_.

Reply:
> don't think the behaviour here necessarily has to be the same as on conversions

I believe it should be the same. The conversion(!) of a `~Value x y z` to real constructor invocation is a kind of conversion - conversion of a _future computation_ described by the `~` and its arguments into real invocation.

Reply:
> API author to never expect sum types that have clashing constructors

Of course. But the problem is that this is out of control of the single API author. As soon as one starts to compose libraries developed by different teams, and let the libraries evolve on their own - clashes will happen. The [Java default methods failure story](http://wiki.apidesign.org/wiki/BinaryCompatibleDefaultMethods) is worth reading.

I am all for **static compilation failing** or generating a warning (when it is on), but **runtime shall try to run** (not crash like in the above [CharSequence.isEmpty case](http://wiki.apidesign.org/wiki/BinaryCompatibleDefaultMethods)). Just like `@Override` annotation would fail `javac` compilation but has no effect on the runtime semantics.

Reply:
Great idea to let [suggestion DB detect](https://github.com/enso-org/enso/issues/9259#issuecomment-1976936469) such overloading and avoid autoscoping in such case.